File: src/lib/mobile/offline-queue.ts
// ═══════════════════════════════════════════════════════════════
// OFFLINE SYNC QUEUE
// ═══════════════════════════════════════════════════════════════
// The hard part of an HR mobile app.
//
// Field staff clock in from basements, warehouses and construction sites where
// there is no signal. An attendance app that requires connectivity to record a
// punch is worse than a paper sheet, because the paper sheet works. So the
// punch is recorded locally and reconciled later.
//
// That makes correctness subtle in ways an online-only app never faces:
//
//  * The punch's timestamp is when it happened, not when it synced. A worker
//    who clocked in at 06:00 and regained signal at 14:00 was not eight hours
//    late.
//  * Delivery is at-least-once, so the server may see the same punch twice.
//    Every operation carries a client-generated idempotency key.
//  * Some failures are permanent. Retrying a 400 forever burns battery and
//    never succeeds; the item has to be quarantined for a human to look at.
//  * Order matters within a stream. A clock-out that arrives before its
//    clock-in produces nonsense.
//
// Storage is injected, so this is testable without a device and works with
// AsyncStorage, SQLite or MMKV.

export type OperationKind =
  | "attendance.clock_in"
  | "attendance.clock_out"
  | "leave.apply"
  | "expense.submit"
  | "profile.update";

export type QueueStatus = "pending" | "in_flight" | "failed" | "quarantined";

export interface QueuedOperation {
  /** Client-generated, stable across retries; the server's idempotency key. */
  id: string;
  kind: OperationKind;
  payload: Record<string, unknown>;
  /** When the user performed the action, not when it is sent. */
  occurredAt: string;
  status: QueueStatus;
  attempts: number;
  lastAttemptAt?: string;
  lastError?: string;
  /** Operations sharing a key are sent in order relative to each other. */
  streamKey?: string;
}

export interface QueueStorage {
  read(): Promise<QueuedOperation[]>;
  write(operations: QueuedOperation[]): Promise<void>;
}

export interface SyncResult {
  ok: boolean;
  /** Permanent failure: do not retry, quarantine instead. */
  permanent?: boolean;
  error?: string;
}

export type Sender = (operation: QueuedOperation) => Promise<SyncResult>;

/** Attempts before an item is quarantined for manual review. */
const MAX_ATTEMPTS = 8;

/**
 * Cap on queue size.
 *
 * A device offline for weeks must not fill its storage. When the cap is hit
 * the oldest *synced-irrelevant* entries go first; pending work is never
 * silently dropped, because that is someone's unpaid hours.
 */
const MAX_QUEUE_SIZE = 500;

/**
 * Exponential backoff with jitter, in milliseconds.
 *
 * Jitter matters more than usual here: when an office regains connectivity,
 * every device retries at once. Identical backoff turns that into a
 * thundering herd against the API.
 */
export function backoffMs(attempts: number, random: () => number = Math.random): number {
  const base = Math.min(2 ** attempts * 1_000, 5 * 60_000);
  return Math.round(base * (0.5 + random() * 0.5));
}

export function isDue(
  operation: QueuedOperation,
  now: number,
  random: () => number = () => 1
): boolean {
  if (operation.status === "quarantined" || operation.status === "in_flight") return false;
  if (!operation.lastAttemptAt) return true;
  return now - new Date(operation.lastAttemptAt).getTime() >= backoffMs(operation.attempts, random);
}

export class OfflineQueue {
  constructor(
    private readonly storage: QueueStorage,
    private readonly send: Sender,
    private readonly now: () => number = Date.now
  ) {}

  /** Adds an operation, recording when it actually happened. */
  async enqueue(
    kind: OperationKind,
    payload: Record<string, unknown>,
    options: { id: string; occurredAt?: string; streamKey?: string }
  ): Promise<QueuedOperation> {
    const operations = await this.storage.read();

    // The same id twice is a double-tap or a re-render, not a second action.
    const existing = operations.find((o) => o.id === options.id);
    if (existing) return existing;

    const operation: QueuedOperation = {
      id: options.id,
      kind,
      payload,
      occurredAt: options.occurredAt ?? new Date(this.now()).toISOString(),
      status: "pending",
      attempts: 0,
      streamKey: options.streamKey,
    };

    const next = [...operations, operation];
    await this.storage.write(this.trim(next));
    return operation;
  }

  /**
   * Drops the oldest resolved entries when over the cap.
   *
   * Quarantined items are kept: they represent work that never reached the
   * server and needs a human decision. Discarding them would lose someone's
   * hours silently, which is the one outcome worse than a full queue.
   */
  private trim(operations: QueuedOperation[]): QueuedOperation[] {
    if (operations.length <= MAX_QUEUE_SIZE) return operations;

    const keep = operations.filter((o) => o.status !== "failed");
    const droppable = operations
      .filter((o) => o.status === "failed")
      .sort((a, b) => a.occurredAt.localeCompare(b.occurredAt));

    const room = Math.max(0, MAX_QUEUE_SIZE - keep.length);
    return [...keep, ...droppable.slice(droppable.length - room)].sort((a, b) =>
      a.occurredAt.localeCompare(b.occurredAt)
    );
  }

  async pending(): Promise<QueuedOperation[]> {
    const operations = await this.storage.read();
    return operations.filter((o) => o.status === "pending" || o.status === "failed");
  }

  async quarantined(): Promise<QueuedOperation[]> {
    const operations = await this.storage.read();
    return operations.filter((o) => o.status === "quarantined");
  }

  /**
   * Sends everything that is due.
   *
   * Operations are processed oldest-first. Within a stream, a failure stops
   * that stream — sending a clock-out after its clock-in failed would record
   * an impossible day — but other streams keep going, so one employee's stuck
   * punch does not block a shared tablet's other users.
   */
  async flush(random: () => number = Math.random): Promise<{
    sent: number;
    failed: number;
    quarantined: number;
  }> {
    const operations = await this.storage.read();
    const now = this.now();

    const due = operations
      .filter((o) => isDue(o, now, random))
      .sort((a, b) => a.occurredAt.localeCompare(b.occurredAt));

    let sent = 0;
    let failed = 0;
    let quarantined = 0;
    const blockedStreams = new Set<string>();
    const resolved = new Set<string>();

    for (const operation of due) {
      if (operation.streamKey && blockedStreams.has(operation.streamKey)) continue;

      let result: SyncResult;
      try {
        result = await this.send(operation);
      } catch (error) {
        // A sender that throws is treated as a transient network failure, not
        // a permanent rejection — the request may never have reached a server.
        result = { ok: false, error: error instanceof Error ? error.message : String(error) };
      }

      const attempts = operation.attempts + 1;

      if (result.ok) {
        resolved.add(operation.id);
        sent++;
        continue;
      }

      const isQuarantine = result.permanent || attempts >= MAX_ATTEMPTS;
      operation.status = isQuarantine ? "quarantined" : "failed";
      operation.attempts = attempts;
      operation.lastAttemptAt = new Date(now).toISOString();
      operation.lastError = result.error;

      if (isQuarantine) quarantined++;
      else failed++;

      if (operation.streamKey) blockedStreams.add(operation.streamKey);
    }

    // Successful operations are removed rather than marked done: the queue is
    // a to-do list, not a history, and history lives on the server.
    await this.storage.write(operations.filter((o) => !resolved.has(o.id)));

    return { sent, failed, quarantined };
  }

  /**
   * What became of a submitted operation.
   *
   * Successful operations are deleted, so absence means sent. That is a
   * conclusion worth stating in one place rather than re-deriving at each
   * call site: the obvious version — "is it still in `pending()`?" — reports
   * a *quarantined* operation as sent, because `pending()` deliberately
   * excludes quarantined work. The caller then tells someone their clock-in
   * was recorded when it was permanently rejected, which is the one thing an
   * attendance app must never say.
   */
  async outcomeOf(id: string): Promise<"sent" | "queued" | "quarantined"> {
    const operations = await this.storage.read();
    const operation = operations.find((o) => o.id === id);

    if (!operation) return "sent";
    return operation.status === "quarantined" ? "quarantined" : "queued";
  }

  /** Returns a quarantined operation to the queue after the user retries it. */
  async retryQuarantined(id: string): Promise<boolean> {
    const operations = await this.storage.read();
    const operation = operations.find((o) => o.id === id && o.status === "quarantined");
    if (!operation) return false;

    operation.status = "pending";
    operation.attempts = 0;
    operation.lastAttemptAt = undefined;
    operation.lastError = undefined;

    await this.storage.write(operations);
    return true;
  }

  /** Discards a quarantined operation. Only ever driven by an explicit action. */
  async discard(id: string): Promise<boolean> {
    const operations = await this.storage.read();
    const next = operations.filter((o) => !(o.id === id && o.status === "quarantined"));
    if (next.length === operations.length) return false;

    await this.storage.write(next);
    return true;
  }

  /** Counts for the sync indicator in the app shell. */
  async status(): Promise<{ pending: number; failed: number; quarantined: number }> {
    const operations = await this.storage.read();
    return {
      pending: operations.filter((o) => o.status === "pending").length,
      failed: operations.filter((o) => o.status === "failed").length,
      quarantined: operations.filter((o) => o.status === "quarantined").length,
    };
  }
}

/**
 * Classifies an HTTP response for retry purposes.
 *
 * The distinction that matters: 4xx means the request will never succeed as
 * written, so retrying wastes battery and hides the problem. 5xx and network
 * errors are worth retrying. 429 is the exception among 4xx — it explicitly
 * asks the client to come back later. 409 is treated as success because the
 * server has already recorded this idempotency key.
 */
export function classifyResponse(status: number): SyncResult {
  if (status >= 200 && status < 300) return { ok: true };
  if (status === 409) return { ok: true };
  if (status === 429) return { ok: false, error: "Rate limited" };
  if (status >= 400 && status < 500) {
    return { ok: false, permanent: true, error: `Rejected with ${status}` };
  }
  return { ok: false, error: `Server error ${status}` };
}

File: android/app/src/main/java/com/circuvent/hrms/data/queue/OfflineQueue.kt
package com.circuvent.hrms.data.queue

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import kotlin.math.min
import kotlin.math.pow

/**
 * THE OFFLINE QUEUE
 *
 * The reason this app exists. Somebody clocking in from a basement car park
 * has no signal, and their punch is the record that they were at work.
 *
 * Written to disk *before* it is sent, not sent-then-queued-on-failure. The
 * process can be killed between the tap and the response — locking the phone
 * and putting it in a pocket does exactly that — and the punch has to survive
 * it.
 *
 * Three properties matter, and each one is a defect that shipped somewhere
 * before it was understood:
 *
 *  * **Idempotency.** Every operation carries a client-generated id that is
 *    stable across retries and is sent as `Idempotency-Key`. A double tap on a
 *    slow network must not produce two punches.
 *  * **Ordering.** Operations on the same `streamKey` are sent in the order
 *    they were made. A clock-out that overtakes its clock-in records an
 *    impossible day.
 *  * **Three outcomes, not two.** Sent, queued, or refused. A permanently
 *    refused punch that reports as "clocked in" is the worst possible result,
 *    because the person stops thinking about it and finds out at payday.
 *
 * Plain SQLite rather than Room: one table, no relations, and Room would bring
 * an annotation processor and a compiler-plugin version to keep in step for a
 * schema that fits on a screen.
 */
class OfflineQueue(context: Context) {

    enum class Status { PENDING, IN_FLIGHT, FAILED, QUARANTINED }

    data class Operation(
        val id: String,
        val kind: String,
        val payload: String,
        val streamKey: String?,
        val status: Status,
        val attempts: Int,
        val nextAttemptAt: Long,
        val createdAt: Long,
        val lastError: String?,
    )

    private val helper = object : SQLiteOpenHelper(context, DB, null, VERSION) {
        override fun onCreate(db: SQLiteDatabase) {
            db.execSQL(
                """
                CREATE TABLE $TABLE (
                    id TEXT PRIMARY KEY NOT NULL,
                    kind TEXT NOT NULL,
                    payload TEXT NOT NULL,
                    stream_key TEXT,
                    status TEXT NOT NULL,
                    attempts INTEGER NOT NULL DEFAULT 0,
                    next_attempt_at INTEGER NOT NULL DEFAULT 0,
                    created_at INTEGER NOT NULL,
                    last_error TEXT
                )
                """.trimIndent()
            )
            // Reading the queue is "what is due, oldest first", on every tick.
            db.execSQL("CREATE INDEX idx_due ON $TABLE (status, next_attempt_at, created_at)")
        }

        override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
            // Never dropped. The table can hold somebody's unsent clock-in, and
            // a migration that recreates it loses a record of them being at
            // work. A future version adds columns; it does not start again.
            throw IllegalStateException(
                "No migration from $oldVersion to $newVersion. The queue may hold unsent work."
            )
        }
    }

    /**
     * Adds an operation, or leaves the existing one alone.
     *
     * `CONFLICT_IGNORE` on the primary key is the idempotency guarantee at the
     * device end: enqueueing the same id twice — which is what a double tap
     * does — is one row.
     */
    fun enqueue(id: String, kind: String, payload: String, streamKey: String?): Boolean {
        val values = ContentValues().apply {
            put("id", id)
            put("kind", kind)
            put("payload", payload)
            put("stream_key", streamKey)
            put("status", Status.PENDING.name)
            put("attempts", 0)
            put("next_attempt_at", 0L)
            put("created_at", System.currentTimeMillis())
        }
        val row = helper.writableDatabase.insertWithOnConflict(
            TABLE, null, values, SQLiteDatabase.CONFLICT_IGNORE
        )
        return row != -1L
    }

    /**
     * What may be sent now.
     *
     * Only the head of each stream is returned. An operation behind a blocked
     * one in the same stream waits, because the order is the point.
     */
    fun due(now: Long = System.currentTimeMillis()): List<Operation> {
        val all = query(
            "status IN (?, ?) AND next_attempt_at <= ?",
            arrayOf(Status.PENDING.name, Status.FAILED.name, now.toString()),
        )

        val seenStreams = mutableSetOf<String>()
        return all.filter { operation ->
            val stream = operation.streamKey ?: return@filter true
            seenStreams.add(stream)
        }
    }

    fun pending(): List<Operation> =
        query("status IN (?, ?)", arrayOf(Status.PENDING.name, Status.FAILED.name))

    fun quarantined(): List<Operation> =
        query("status = ?", arrayOf(Status.QUARANTINED.name))

    /**
     * The outcome of one operation, in three states.
     *
     * Not a boolean. `submit()` in the previous implementation asked "is this
     * still pending?" to decide whether it had worked — and `pending()`
     * deliberately excludes quarantined work, so a permanently refused punch
     * came back as success. The clock-in screen said "Clocked in" to somebody
     * whose punch the server had rejected outright.
     */
    fun outcomeOf(id: String): Status? {
        helper.readableDatabase.query(
            TABLE, arrayOf("status"), "id = ?", arrayOf(id), null, null, null
        ).use { cursor ->
            if (!cursor.moveToFirst()) return null
            return Status.valueOf(cursor.getString(0))
        }
    }

    fun markSent(id: String) {
        helper.writableDatabase.delete(TABLE, "id = ?", arrayOf(id))
    }

    /**
     * Records a failure and decides whether it is worth trying again.
     *
     * A 4xx other than 408 or 429 is the server saying no on the merits: the
     * payload is wrong, or the action is not allowed. Retrying it forever
     * would hide a refusal behind an amber "sending" state for ever. Anything
     * else — a 5xx, a timeout, a dead network — is retried with exponential
     * backoff.
     */
    fun markFailed(id: String, status: Int?, error: String?, now: Long = System.currentTimeMillis()) {
        val permanent = status != null && status in 400..499 && status != 408 && status != 429
        val db = helper.writableDatabase

        if (permanent) {
            db.execSQL(
                "UPDATE $TABLE SET status = ?, last_error = ?, attempts = attempts + 1 WHERE id = ?",
                arrayOf(Status.QUARANTINED.name, error, id),
            )
            return
        }

        db.execSQL(
            "UPDATE $TABLE SET status = ?, last_error = ?, attempts = attempts + 1, " +
                "next_attempt_at = ? WHERE id = ?",
            arrayOf(Status.FAILED.name, error, backoffFrom(id, now), id),
        )
    }

    private fun backoffFrom(id: String, now: Long): Long {
        val attempts = query("id = ?", arrayOf(id)).firstOrNull()?.attempts ?: 0
        return now + backoffMs(attempts + 1)
    }

    /** Puts refused work back in the queue, only ever at the user's request. */
    fun retry(id: String) {
        helper.writableDatabase.execSQL(
            "UPDATE $TABLE SET status = ?, next_attempt_at = 0, last_error = NULL WHERE id = ?",
            arrayOf(Status.PENDING.name, id),
        )
    }

    /** Throws refused work away. Only ever an explicit user action. */
    fun discard(id: String) {
        helper.writableDatabase.delete(TABLE, "id = ?", arrayOf(id))
    }

    private fun query(where: String, args: Array<String>): List<Operation> {
        helper.readableDatabase.query(
            TABLE, null, where, args, null, null, "created_at ASC"
        ).use { cursor ->
            val out = mutableListOf<Operation>()
            while (cursor.moveToNext()) {
                out += Operation(
                    id = cursor.getString(cursor.getColumnIndexOrThrow("id")),
                    kind = cursor.getString(cursor.getColumnIndexOrThrow("kind")),
                    payload = cursor.getString(cursor.getColumnIndexOrThrow("payload")),
                    streamKey = cursor.getString(cursor.getColumnIndexOrThrow("stream_key")),
                    status = Status.valueOf(cursor.getString(cursor.getColumnIndexOrThrow("status"))),
                    attempts = cursor.getInt(cursor.getColumnIndexOrThrow("attempts")),
                    nextAttemptAt = cursor.getLong(cursor.getColumnIndexOrThrow("next_attempt_at")),
                    createdAt = cursor.getLong(cursor.getColumnIndexOrThrow("created_at")),
                    lastError = cursor.getString(cursor.getColumnIndexOrThrow("last_error")),
                )
            }
            return out
        }
    }

    companion object {
        private const val DB = "circuvent_queue.db"
        private const val VERSION = 1
        private const val TABLE = "operations"

        /**
         * Exponential backoff, capped at five minutes.
         *
         * Capped because this runs while the app is open and somebody is
         * looking at it. An hour-long backoff on a phone that regained signal
         * two minutes ago means a punch that sits there for no reason.
         */
        fun backoffMs(attempt: Int): Long {
            if (attempt <= 0) return 0
            val exponential = 1_000L * 2.0.pow(attempt - 1).toLong()
            return min(exponential, 5 * 60 * 1000L)
        }
    }
}
